### PR TITLE
feat: add pm caching

### PR DESCRIPTION
**Pull request by codding-y**

### CODE REVIEW

File: src/BasicAbility/Masa.Contrib.BasicAbility.Pm/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Masa.Contrib.BasicAbility.Pm
+{
+    public class Constants
+    {
+        internal const string DEFAULT_CLIENT_NAME = "Masa.Contrib.BasicAbility.Pm";
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Pm/Constants.cs:11]: Pascal or lower?

File: src/BasicAbility/Masa.Contrib.BasicAbility.Pm/PmCaching.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            else
+            {
+                var projects = await _memoryCacheClient.GetListAsync<Project>(keys.ToArray());
+                var result = projects.Where(project => project != null).ToList();
+
+                return result!;
+            }
+        }
+
+        public async Task<List<App>> GetAppListAsync(string envName)
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Pm/PmCaching.cs:31]: GetAppListAsync means GetAppListByProjectAsync, maybe can call `GetAllAppsAsync`, and why need this api?

File: test/Masa.Contrib.BasicAbility.Pm.Tests/PmCachingTest.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                redisOptions.Servers = new List<RedisServerOptions>
+                {
+                    new RedisServerOptions("10.10.90.37",30120)
+                };
+                redisOptions.DefaultDatabase = 0;
+                redisOptions.Password = "p@ssw0rd";
+            }, callerOptions =>
+            {
+                callerOptions.UseHttpClient(builder =>
```
Comment (doddgu) [inline test/Masa.Contrib.BasicAbility.Pm.Tests/PmCachingTest.cs:33]: Why need this caller?

File: test/Masa.Contrib.BasicAbility.Pm.Tests/PmCachingTest.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        [TestInitialize]
+        public void Initialize()
+        {
+            var services = new ServiceCollection();
+
+            services.AddPmCaching(redisOptions =>
+            {
+                redisOptions.Servers = new List<RedisServerOptions>
+                {
+                    new RedisServerOptions("10.10.90.37",30120)
```
Comment (zhenlei520) [inline test/Masa.Contrib.BasicAbility.Pm.Tests/PmCachingTest.cs:27]: real redis address?

File: src/BasicAbility/Masa.Contrib.BasicAbility.Pm/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+using Masa.Utils.Caching.Redis.DependencyInjection;
+using Masa.Utils.Caching.Redis.Models;
+using Masa.Utils.Caller.HttpClient;
+using Masa.Utils.Configuration.Json;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (zhenlei520) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Pm/ServiceCollectionExtensions.cs:12]: Use global Using

File: src/BasicAbility/Masa.Contrib.BasicAbility.Pm/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace Masa.Contrib.BasicAbility.Pm
+{
```
Comment (zhenlei520) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Pm/ServiceCollectionExtensions.cs:15]: add after namespace; no longer use {}

File: src/BasicAbility/Masa.Contrib.BasicAbility.Pm/Service/ProjectService.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        private readonly ICallerProvider _callerProvider;
+
+        public ProjectService(ICallerProvider callerProvider)
+        {
+            _callerProvider = callerProvider;
+        }
+
+        public async Task<List<ProjectModel>> GetProjectListAsync(string envName)
+        {
+            var requestUri = $"api/v1/project-with-apps/{envName}";
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Pm/Service/ProjectService.cs:15]: https://www.sitepoint.com/rest-api/

**Pull request closed without merging**